### PR TITLE
Target kotlin 1.4 in kotlin extension

### PR DESCRIPTION
**Pull request by laurit**
Resolves https://github.com/open-telemetry/opentelemetry-java/issues/5903
1.4 should be the earliest supported version for 1.9 compiler, probably we'll have to bump this with the next kotlin version

### CODE REVIEW

File: extensions/kotlin/build.gradle.kts
Comment (laurit) [inline extensions/kotlin/build.gradle.kts:52]: I think this also applies to the tests. Could also use
```
compileKotlin {
  kotlinOptions {
    languageVersion = "1.4"
  }
}
```

**Review by jack-berg — APPROVED** [review]
This looks good to me, but wondering if folks with more kotlin experience like @breedx-splk / @jkwatson / @likethesalad can chime in to confirm.

**Comment by LikeTheSalad** [comment]
> This looks good to me, but wondering if folks with more kotlin experience like @breedx-splk / @jkwatson / @LikeTheSalad can chime in to confirm.

I was actually having issues with this lib forcing me to upgrade Kotlin a couple of days ago though I could sort it out by upgrading to Kotlin 1.8, I tried 1.9 first but that caused a bunch of other issues with other dependencies and it was a mess.. Luckily 1.8 worked out well, still I'm glad this change is being added because I understand that upgrading Kotlin can be painful sometimes. Tbh I've never used the config `languageVersion` before, but it seems the right approach without having to downgrade the Kotlin version used in this project.

> 1.4 should be the earliest supported version for 1.9 compiler, probably we'll have to bump this with the next kotlin version

I think 1.4 is a great option. In the Android gradle plugin they made 1.5 a requirement [a while ago](https://developer.android.com/build/releases/past-releases/agp-7-3-0-release-notes#min-kotlin-version),  so I wouldn't mind bumping it up in the next kotlin version in case 1.4 is no longer supported by then.

**Comment by jkwatson** [comment]
> This looks good to me, but wondering if folks with more kotlin experience like @breedx-splk / @jkwatson / @LikeTheSalad can chime in to confirm.

I have very little kotlin experience, so I'll leave this to the experts.

**Comment by trask** [comment]
> I think 1.4 is a great option. In the Android gradle plugin they made 1.5 a requirement [a while ago](https://developer.android.com/build/releases/past-releases/agp-7-3-0-release-notes#min-kotlin-version), so I wouldn't mind bumping it up in the next kotlin version in case 1.4 is no longer supported by then.

since this artifact is marked stable, would dropping support for older kotlin version be considered a breaking change? (in the instrumentation repo this is why we encode the base version into the artifact name)

**Comment by LikeTheSalad** [comment]
> since this artifact is marked stable, would dropping support for older kotlin version be considered a breaking change?

I think so, because in my case my project stopped compiling until I bumped its Kotlin version. However, it seems like Kotlin is dropping support for older language versions with every new minor version bump release, so it seems unavoidable to drop support for 1.4 at some point, unless we pin this project's Kotlin version to the last one that supports 1.4.

**Comment by jack-berg** [comment]
@trask are you suggesting that this PR is a breaking change or that upgrading to 1.5 after merging this PR would be a breaking change?

**Comment by trask** [comment]
> @trask are you suggesting that this PR is a breaking change or that upgrading to 1.5 after merging this PR would be a breaking change?

the latter, and this was what got my attention:

> probably we'll have to bump this with the next kotlin version

**Comment by jack-berg** [comment]
Hmm.. so the question is should we do something now do accommodate upgrading the minimum kotlin version (or at least have a plan for how we would handle it), or should we codify that we don't consider upgrading the kotlin version a breaking change.

Was looking at the [okhttp changelog](https://square.github.io/okhttp/changelogs/changelog_4x/) the other day and noticed that they upgrade the kotlin version in minor release versions.

**Comment by jack-berg** [comment]
This is related to the minimum android conversation in #5936. I think we should codify that upgrading the minimum android version should not require a new major version, and that the same advice applies to kotlin. Would be good to define some logic for how we choose the minimum kotlin version and document it.

**Comment by LikeTheSalad** [comment]
> Would be good to define some logic for how we choose the minimum kotlin version and document it.

For Android purposes, I believe a good measure could be the [AndroidX libs](https://github.com/androidx/androidx). They are the official Android tools outside of the Android SDK.

They are also the reason why many times I've had to upgrade my kotlin version after upgrading one of them due to compilation issues caused by kotlin version conflicts.

It seems like they've recently upgraded to kotlin `1.9`, and set the [languageVersion from 1.5 to 1.8](https://github.com/androidx/androidx/commit/aa9dff9b6787a50db10504c1f6b1ba5e07e7abff). I'm not sure why they've jumped straight to `1.8` though, probably there were not many breaking changes from `1.5` up to `1.8`... Though I'm not fully sure. In any case, I think this repo should be a good guide on what kotlin version to use when it comes to Android projects, not sure about non-android projects that use kotlin though.

One thing to keep an eye on is that this repo seems to set the `languageVersion` in two places, one for their build tools [here](https://github.com/androidx/androidx/blob/fba9976441712daa63374c9595dd7792a613454f/buildSrc/shared.gradle#L42) and one for their deliverable projects [here](https://github.com/androidx/androidx/blob/92416107b23adff21b1b19a64a2aa2e1b46583b4/buildSrc/public/src/main/kotlin/androidx/build/AndroidXConfiguration.kt#L44) (which is later set to subprojects [here](https://github.com/androidx/androidx/blob/androidx-main/buildSrc/private/src/main/kotlin/androidx/build/AndroidXImplPlugin.kt#L344C75-L344C75)). They're both set to `1.8` atm so it's not a problem, but it's probably worth taking a look at both once in a while.

**Comment by jack-berg** [comment]
@LikeTheSalad I think upgrading to 1.9 would conflict with the original issue #5903 in which @laurit describes that Spring boot 2.7.x users require 1.6.21.

Although, after writing that, I'm not sure why a spring boot 2.7.x user couldn't upgrade kotlin to a later version than 1.6.21. Is kotlin 1.9 not compatible with 1.6.21?

**Comment by LikeTheSalad** [comment]
> @LikeTheSalad I think upgrading to 1.9 would conflict with the original issue #5903 in which @laurit describes that Spring boot 2.7.x users require 1.6.21.

Yeah definitely, I also think we shouldn't force `1.9` compatibility (which is the case right now), and the AndroidX lib I mentioned earlier doesn't force it either because they set their `languageVersion` compilation param to `1.8` which ensures compatibility with versions older than `1.9`. So my comment above was more about what's the lowest `languageVersion` we could set to keep compatibility with older kotlin versions, and at least for Android, it seems like it could be `1.8` right now based on AndroidX libs, although that still seems pretty high.

Maybe just a simpler option is to always keep the `languageVersion` to the minimum non-deprecated one, which seems to be [1.6 right now](https://kotlinlang.org/docs/gradle-compiler-options.html#attributes-common-to-jvm-and-js). It's probably what @laurit was trying to do anyway, to use the minimum compat version supported.

> Is kotlin 1.9 not compatible with 1.6.21?

No, I just tried it locally and, if I set my kotlin version to `1.6.21` while I have a dependency built with kotlin `1.9.x`, I get this compilation error message:

`Module was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.9.0, expected version is 1.6.0.`

I also tried with `1.7.x` with the same results. Only `1.8.x` seems to be compatible with `1.9.x` for some reason.

**Comment by laurit** [comment]
> Only 1.8.x seems to be compatible with 1.9.x for some reason.

As far as I understand kotlin compiler accepts metadata that is at most one version ahead of its version.

> Although, after writing that, I'm not sure why a spring boot 2.7.x user couldn't upgrade kotlin to a later version than 1.6.21. Is kotlin 1.9 not compatible with 1.6.21?

https://github.com/open-telemetry/opentelemetry-java/issues/5903#issue-1937501993 states that

> Another alternative could be to use a newer Kotlin version than 1.6.21 myself but I don't know which consequences that would have.

Our choices are to choose the minimum kotlin version which we wish to support and avoid updating kotlin dependency for this module or keep updating the kotlin dependency and set the target to earliest supported version. Either is better than what we do now.

**Comment by jack-berg** [comment]
> or keep updating the kotlin dependency and set the target to earliest supported version

Let's do this, and document this strategy in [versioning](https://github.com/open-telemetry/opentelemetry-java/blob/main/VERSIONING.md). Avoiding updating the kotlin dependency is too restrictive without any strong justification. I'd like to retain the flexibility to update the kotlin version in the future.

File: extensions/kotlin/build.gradle.kts
Comment (jack-berg) [inline extensions/kotlin/build.gradle.kts:52]: #5983 defines our minimum kotlin language version to be 1.6. 
```suggestion
      languageVersion = "1.6"
```

**Review by jack-berg — APPROVED** [review]

**Pull request merged**
